kumascript/macros/CSSSyntax.ejs

```diff
@@ -1,442 +1,324 @@
 }
 
-function addBrackets(str) {
-    return str.match(/\(\)$/) ? str : str + "()";
+/**
+ * Determines the markup to generate for a single node in the AST
+ * generated by css-tree.
+ */
+function renderNode(name, node) {
+  switch (node.type) {
+    case 'Property': {
+      return `<span class="token property">${name}</span>`;
+    }
+    case 'Type': {
+      // encode < and >
+      let encoded = replaceAll(name, '<', '&lt;');
+      encoded = replaceAll(encoded, '>', '&gt;');
+      // add CSS class: we use "property" because there isn't one for types
+      const span = `<span class="token property">${encoded}</span>`;
+      // if this type is not included in the syntax, link to its dedicated page
+      if (valuespaces[name] && valuespaces[name].value) {
+        return span;
+      } else {
+        // the slug does not include the angle brackets
+        let slug = replaceAll(name, '<', '');
+        slug = replaceAll(slug, '>', '');
+        slug = slug.replace(/\[.*\]/, '')
+        return `<a href="/${locale}/docs/Web/CSS/${slug}">${span}</a>`;
+      }
+    }
+    case 'Multiplier': {
+      // link to the Value Definition Symtax and provide a tooltip
+      let key = name;
+      if (name.startsWith('{')) {
+        key = '{}';
+      }
+      const info = syntaxDescriptions[key];
+      return `<a href="${valueDefinitionUrl}#${info.fragment}" title="${info.tooltip}">${name}</a>`;
+    }
+    case 'Keyword': {
+      return `<span class="token keyword">${name}</span>`;
+    }
+    case 'Function': {
+      return `<span class="token function">${name}</span>`;
+    }
+    case 'Token': {
+      if (name === ')') {
+        // this is probably a closing bracket
+        return `<span class="token function">${name}</span>`;
+      }
+    }
+    case 'Group': {
+      // link from brackets to the value definition syntax docs
+      const info = syntaxDescriptions['[]'];
+      name = name.replace(/^\[/, `<a href="${valueDefinitionUrl}#${info.fragment}" title="${info.tooltip}">[</a>`);
+      name = name.replace(/\]$/, `<a href="${valueDefinitionUrl}#${info.fragment}" title="${info.tooltip}">]</a>`);
+
+      // link from combinators (except " ") to the value definition syntax docs
+      if (node.combinator && (node.combinator !== ' ')) {
+        const info = syntaxDescriptions[node.combinator];
+        // note that we are replacing the combinator surrounded by spaces, like " | "
+        name = replaceAll(name, ` ${node.combinator} `, ` <a href="${valueDefinitionUrl}#${info.fragment}" title="${info.tooltip}">${node.combinator}</a> `);
+      }
+
+      return name;
+    }
+  }
+  return name;
 }
 
-async function formatSyntax(rawSyntax) {
-    if (rawSyntax === "") {
-        return "";
+/**
+ * Generate the markup for every term in a syntax definition,
+ * ensuring that the terms are visually aligned
+ */
+function renderTerms(terms, combinator) {
+  let output = '';
+  const termArray = [];
+  const termTextLengths = [];
```

Comment:
Would be nice to combine these into a single array, which holds objects of type `{ length: number, text: string }`.

Btw `let maxTermLength = Math.max(...termTextLengths);` becomes `const maxTermLength = Math.max(...renderedTerms.map(t => t.length));`.

Reply:
yeah, this is nice :)